Lootrun command and path rendering

kristofbolyai opened the pull request:
This PR is porting @coehlrich's Lootrun path renderer into Artemis. 

Mod loader versions are also updated. (Mostly unrelated to PR, at some point I thought the currently used Forge version was buggy.)

HighCrit commented:
It'll likely not fix the problem only prevent crashes

kristofbolyai committed:
Add access wideners

kristofbolyai committed:
Update forge

kristofbolyai committed:
Update parchment

kristofbolyai committed:
Update fabric

kristofbolyai committed:
Add classes related to lootruns

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai committed:
Added lots of lootrun related classes

kristofbolyai committed:
Use MCUtils

kristofbolyai committed:
Only have LootrunUtils events registered when lootrun is loaded

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai committed:
Started adding translations

kristofbolyai committed:
Finished adding translations

kristofbolyai committed:
Cleanup code

ryanzhoudev commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,544 @@
+            int cycle = 10 * cycleDistance;
+            int parts = i % cycle;
+            float done = (float) parts / (float) cycle;
+            int usedColor = 0;
```
usedColor is immediately set to currentColor in both the if and else conditions

ryanzhoudev commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,427 @@
+/*
+ * Copyright © Wynntils 2021.
```
Update all of these to 2022? I think there are a bunch of other files that also say 2021 and I'm guessing most of us just copy those into new files

ryanzhoudev commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -17,5 +17,33 @@
+  "feature.wynntils.lootrunUtils.addedNote": "Added note at %s with text: ",
+  "feature.wynntils.lootrunUtils.savedLootrun": "Successfully saved lootrun",
+  "feature.wynntils.lootrunUtils.errorSavingLootrun": "Error saving lootrun",
+  "feature.wynntils.lootrunUtils.errorSavingLootrunAlreadyExists": "A lootrun with this name already exists,",
```
Trailing comma

ryanzhoudev commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -17,5 +17,33 @@
-  "featureDebug.wynntils.packetDebugger.name": "Packet Debugger Feature"
+  "featureDebug.wynntils.packetDebugger.name": "Packet Debugger Feature",
+  "feature.wynntils.lootrunUtils.recordStart":"Started recording lootrun. Execute %s to stop recording.",
+  "feature.wynntils.lootrunUtils.recordStop1":"Stopped recording lootrun. Execute %s to delete the lootrun",
```
Small thing; maybe run over this file and add applicable ending punctuation for consistency?

ryanzhoudev commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,544 @@
+
+public class LootrunUtils {
+
+    public static final File LOOTRUNS = new File(McUtils.mc().gameDirectory, "lootruns");
```
This folder is not under the `wynntils` subfolder where `apicache` currently sits, is this intentional?

ryanzhoudev commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,427 @@
+
+    @Override
+    public void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        dispatcher.register(Commands.literal("lootrun")
```
Could you add an alias to "lr" like Wynntils 1.12.2? I can do it if you don't want to

ryanzhoudev commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,544 @@
+        renderLootrun(event, recordingCompiled, RECORDING_COLOR);
+    }
+
+    private static void renderLootrun(RenderLevelLastEvent event, LootrunInstance lootrun, int color) {
```
The rendering for lootrun paths seems kind of broken - sometimes it'll stop rendering a certain portion when I move around
![image](https://user-images.githubusercontent.com/57310593/167705574-076657a9-4a07-4f1d-a056-f7d5f60a5750.png)
![image](https://user-images.githubusercontent.com/57310593/167705612-8333f369-b426-4d3c-9eb5-a3601e3f12e1.png)
Unless this is intentional and you only want to load portions closest to the player?

kristofbolyai replied:
I am not sure about this, I will check later but maybe coehlrich knows better

kristofbolyai replied:
Also, I can't seem to reproduce this, can you maybe share the affected file here?

kristofbolyai replied:
@DonkeyBlaster you had a low render distance when this happened? If so, it might be related to valid block calculation math. If not, I have no idea.

ryanzhoudev replied:
Yeah, my laptop has a render setting of 2 chunks

kristofbolyai replied:
I checked the code and render distance should not affect this.

kristofbolyai replied:
Fixed this

kristofbolyai committed:
Add lr alias

kristofbolyai committed:
Use mod storage root in LootrunUtils

kristofbolyai committed:
Fix translation punctuations

kristofbolyai committed:
Use correct dates in copyrights

kristofbolyai committed:
Remove unused code

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai committed:
Merge branch 'main' into lootrun

magicus commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,430 @@
+        return 1;
+    }
+
+    private int saveLootrun(CommandContext<CommandSourceStack> context) {
```
Seems like most of this file should be in a Feature, rather than in a Command, yes?

P0keDev replied:
honestly i think a lot of the stuff in the command class should go in `LootrunUtils`, and a lot of stuff currently in `LootrunUtils` could go into a `LootrunFeature` class, to align with our current convention of doing event handling in Features

kristofbolyai replied:
I am not sure what a Lootrun "feature" would do. I always saw commands as "features", as they are separated well enough. Yes, the Util class has SubscribeEvents, but they are only registered if the user turned on recording or loaded a lootrun file. I always imagined features as options that the user can turn on. This time, they are turned on by commands. I don't see how the user would want to disable the `LootrunFeature` as it does not do anything unless turned on by a command.

If you have something in your mind regarding how the `LootrunFeature` would look like, I am happy to discuss it.

P0keDev replied:
a `LootrunFeature` wouldn't be too substantial, but i do think there's value in keeping our event handling standardized - even if it's just calling util methods from the event methods. additionally, if we make our configuration framework group by Feature classes (annotated variables, maybe) it would be best to have a place for lootrun options to go

magicus replied:
For me the Commands should be just entries into functionality, a conversion layer between the chat prompt and the functionality.

See it like this. If we were to add a button to the ingame menu "Start lootrun", should that call into a Command class? That seems wrong. (In fact, I think we have some kind of load lootrun selection GUI in legacy..?)

Once again, I feel we're missing a place to put "Wynntils functionality", that is not a direct part of a feature. Could be have a `functionality` package, as a peer to `features`, for Wynntils functionality that is not bundled as a feature to be turned on and off? Otherwise I think the best solution still is to shoehorn the lootruns into a Feature.

magicus replied:
And interesting idea with Features turned off and on by command. What if we had a `LootrunManagerFeature` that handles lootrun management (loading, saving, deleting and --which is missing in legacy -- editing), and a `LootrunRecording` and `LootrunPlayback`/`LootrunDisplaying` that are enabled, temporary, when saving or playing back/showing a lootrun?

Would that be a good idea, or just confuse things?

kristofbolyai replied:
I feel like the `LootrunManagerFeature` is not really a feature, cause it does not do anything by itself. It can not be "enabled" or "disabled", I don't feel like there is anything to disable.

As for `LootrunRecordingFeature` and `LootrunPlaybackFeature`, they do feel more connected to features, but then they are not really like any other feature, where it is clear what enabling and disabling them does. 

I think it may be the best option to invent something like the aforementioned "functionality", although I don't have solid ideas on how it would look like.

magicus commented on common/src/main/java/com/wynntils/core/WynntilsMod.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright © Wynntils 2021.
```
Spurious change of year

magicus commented on common/src/main/java/com/wynntils/mc/mixin/LevelRendererMixin.java:
```diff
@@ -0,0 +1,43 @@
+    private Minecraft minecraft;
+
+    @Inject(at = @At("TAIL"), method = "renderLevel")
+    public void renderLevel(
```
Should be `renderLevelPost` to follow our naming conventions.

magicus commented on common/src/main/java/com/wynntils/mc/mixin/MultiPlayerGameModeMixin.java:
```diff
@@ -30,4 +37,14 @@ private void handleInventoryMouseClickPre(
     }
+
+    @Inject(method = "useItemOn", at = @At("HEAD"))
+    private void useItemOn(
```
Should be `useItemOnPre`.

magicus commented on common/src/main/java/com/wynntils/mc/mixin/ServerPlayerGameModeMixin.java:
```diff
@@ -0,0 +1,38 @@
+    protected ServerPlayer player;
+
+    @Inject(method = "useItemOn", at = @At("HEAD"))
+    private void useItemOn(
```
`useItemOnPre`

magicus commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,542 @@
+                                for (BlockPos blockInArea : blocks) {
+                                    BlockState blockStateInArea = level.getBlockState(blockInArea);
+                                    if (blockStateInArea.is(Blocks.BARRIER)) {
+                                        // System.out.println("barrier");
```
Remove debug print

magicus commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,542 @@
+                                        .color(pathColor)
+                                        .normal(0, 0, 1)
+                                        .endVertex();
+                                // GLX._renderCrosshair(color, pauseDraw, disabled, disable);
```
Commented-out code should be re-enabled if needed, or deleted.

magicus commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,542 @@
+        renderLootrun(event, recordingCompiled, RECORDING_COLOR);
+    }
+
+    private static void renderLootrun(RenderLevelLastEvent event, LootrunInstance lootrun, int color) {
```
This entire rendering function is not a thing of beauty. :( It is hard to read, no comments, way to long, and uses several booleans affecting flow, which makes it even harder to track.

At the very least, extract the points, chest and notes rendering parts to separate functions. After that, the points rendering will probably need a pass or two of cleanup to make it comprehensible. It seem to be doing some kind of "valid block" checking that can possibly be extracted in a separate function as well.

A simple rule of thumb is that a function should be at most a screen long. This is not always possible to keep, but this method is 150+ lines, which is way to much.

magicus commented on common/src/main/java/com/wynntils/wc/utils/lootrun/objects/Point.java:
```diff
@@ -0,0 +1,9 @@
+
+import net.minecraft.world.phys.Vec3;
+
+public record Point(Vec3 vec3, int color) {}
```
Maybe `ColoredPoint`?

kristofbolyai replied:
Renamed

magicus commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,430 @@
+                                                            .withStyle(ChatFormatting.UNDERLINE)
+                                                            .withStyle((style) -> style.withClickEvent(new ClickEvent(
+                                                                    ClickEvent.Action.SUGGEST_COMMAND,
+                                                                    "/lootrun save "))))
```
This extra space should not be there.

kristofbolyai replied:
Why? It is meant to be a command suggestion and the command has name as an argument. So it looks like `/lootrun save <name>`

magicus replied:
Then I misread the formatting string. Even so, would it not be better to have the space in the surrounding localized string?

kristofbolyai replied:
I think you might have missed that the ClickEvent is a SUGGEST_COMMMAND which opens chat and basically puts "/lootrun save ". The space is there so the player only has to type the name

magicus replied:
You are absolutely right. I missed that. Now it makes sense. Thanks, and sorry :)

magicus commented:
Overall I think this looks a bit unstructured. How much of this is grabbed directly from the legacy wynntils, and how much is rewritten for Artemis? I remember that I thought the old lootrun code was quite nice. Either I've forgotten how it really looks, or we should try to keep more of the old quality, if possible.

P0keDev commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,542 @@
+                if (points.containsKey(chunkLong)) {
+                    List<List<Point>> locations = points.get(chunkLong);
+                    for (List<Point> locationsInRoute : locations) {
+                        VertexConsumer consumer = source.getBuffer(LootrunRenderType.LOOTRUN_LINE);
```
is it really worth creating a whole `RenderType` class just to use it here? the same thing can always be accomplished by calling `Tesselator.getInstance().getBufferBuilder()` and setting the appropriate options. i definitely agree with magicus that this whole function should be split into helpers

kristofbolyai replied:
I think we can keep this, there is not really a harm, and maybe this would be useful when we add the textured version, If any of y'all have a strong opinion against, then let's remove it.

P0keDev commented on common/src/main/java/com/wynntils/wc/utils/lootrun/objects/LootrunUncompiled.java:
```diff
@@ -0,0 +1,20 @@
+import net.minecraft.network.chat.Component;
+import net.minecraft.world.phys.Vec3;
+
+public record LootrunUncompiled(List<Vec3> points, Set<BlockPos> chests, List<Pair<Vec3, Component>> notes, File file) {
```
since this and `LootrunInstance` are so small and only used within `LootrunUtils`, it might be better to declare them as inner classes instead of dedicating 2 files for their collective ~5 lines of substance

P0keDev replied:
ah, hadn't taken a look at the command class before writing this - but since i think the command class _shouldn't_ be handling these in the first place, it still applies

kristofbolyai replied:
Moved almost all classes, except `LootrunUncompiled`. That class is used in commands as well and there is not a clean way to not really use them.

P0keDev replied:
left some comments in the command class on how i think it could be done. my philosophy behind a util class is that any data handling should be done entirely within that class - having to extract data and modify it from outside the class isn't great coding practice and ideally shouldn't even be possible

magicus replied:
"data encapsulation". Yes, it's a generally good coding style to get the right level of coupling. (I have not looked at the latest changes wrt this in this PR yet.)

But. To me a "util" class is more of a "toolbox" kind of thing. The prototypical Util class contains a number of vaguely related, static, utility methods, with no local data. Like `java.util.Objects`. 

That's actually one thing with how the design has shifted in Artemis that I'm not at all fond of -- it seems that too much domain knowledge has been put under the `util` package. :-( That was never my intention in the original design sketch.

P0keDev commented on common/src/main/java/com/wynntils/mc/mixin/ServerPlayerGameModeMixin.java:
```diff
@@ -0,0 +1,38 @@
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+@Mixin(ServerPlayerGameMode.class)
+public class ServerPlayerGameModeMixin {
```
haven't looked into the vanilla source myself yet, but is there a reason two different mixins are required to handle right clicks? is there no target that could handle everything on its own?

Incompleteusern replied:
yeah ServerPlayerGameMode seems to be server side

kristofbolyai replied:
Yeah, it was a server side thing, removed it.

P0keDev commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,430 @@
+import net.minecraft.network.chat.*;
+import net.minecraft.world.phys.Vec3;
+
+public class LootrunCommand extends CommandBase {
```
it doesn't seem logical to me that so much of the lootrun logic is handled here, especially the way variables in `LootrunUtils` are being directly assigned. if we have a utils class, i feel like we should be doing all that logic inside of it via function calls instead of this

kristofbolyai replied:
Refactored this code, hope its better now

P0keDev commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -18,5 +18,33 @@
   "featureDebug.wynntils.keyBindTest.name": "Connection Progress Feature",
-  "featureDebug.wynntils.packetDebugger.name": "Packet Debugger Feature"
+  "featureDebug.wynntils.packetDebugger.name": "Packet Debugger Feature",
+  "feature.wynntils.lootrunUtils.recordStart":"Started recording lootrun. Execute %s to stop recording.",
```
shouldn't these go above the debug strings?

i feel like it would be good to have some convention for organizing stuff in this file, too, like alphabetical order. can we configure the formatter to do that automatically?

magicus replied:
Alphabetic order sounds good. I don't think we can get this just by setting a switch, but we can probably hook in our own formatter to do this.

kristofbolyai replied:
#58

P0keDev reviewed (commented):
it doesn't need to be addressed in the current PR, but i definitely think we should support a textured lootrun path as well - a simple thin line can be hard to spot, especially if you're trying to move as quickly as possible, and it also loses the directional information

kristofbolyai commented:
> Overall I think this looks a bit unstructured. How much of this is grabbed directly from the legacy wynntils, and how much is rewritten for Artemis? I remember that I thought the old lootrun code was quite nice. Either I've forgotten how it really looks, or we should try to keep more of the old quality, if possible.

I am not sure, as I based this mostly on coehlrich's code, with reworks to fit more into our structure. I will check out the legacy code, then rework this PR.

kristofbolyai committed:
Fix copyright date

kristofbolyai committed:
Remove unnecessary mixin

kristofbolyai committed:
Change method naming to match conventions

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai committed:
Refactored LootrunCommand

kristofbolyai committed:
Merge branch 'lootrun' of https://github.com/kristofbolyai/Artemis into lootrun

kristofbolyai committed:
spotless

kristofbolyai committed:
Moved classes to LootrunUtils

kristofbolyai committed:
Remove debug comments

kristofbolyai committed:
Fix states

P0keDev commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,404 @@
+    }
+
+    private int listLootrunNote(CommandContext<CommandSourceStack> context) {
+        LootrunUncompiled current = LootrunUtils.getActiveLootrun();
```
it would make more sense to replace this with a `getNotes` function, no? with a similar sort of null check

P0keDev commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,404 @@
+    }
+
+    private int deleteLootrunNote(CommandContext<CommandSourceStack> context) throws CommandSyntaxException {
+        LootrunUncompiled current = LootrunUtils.getActiveLootrun();
```
same here - a `deleteNote` function that returns a boolean for success would be cleaner

P0keDev commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,404 @@
+    private int addChest(CommandContext<CommandSourceStack> context) throws CommandSyntaxException {
+        BlockPos pos = BlockPosArgument.getSpawnablePos(context, "pos");
+
+        LootrunUncompiled current = LootrunUtils.getActiveLootrun();
```
similarly, `addChest` and `removeChest` for this method and the one below

kristofbolyai committed:
Refactor rendering in LootrunUtils

P0keDev commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,404 @@
+            return 0;
+        } else {
+            Vec3 position = McUtils.player().position();
+            List<Vec3> points = LootrunUtils.getRecording().points();
```
could be an `undoToPos` method or something similar

kristofbolyai committed:
Reworked command, moved all logic to LootrunUtils

kristofbolyai committed:
Moved LootrunUncompiled

kristofbolyai commented:
@magicus @P0keDev @Incompleteusern I did the first round of cleanups, refactored a few things. Cleaned up the rendering, it's much better now imo. I also reworked commands to have all logic moved to `LootrunUtils`.

The last thing, I will move event logic to a `LootrunFeature` class tomorrow, don't have time for it now.

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai committed:
Moved events to LootrunFeature

kristofbolyai committed:
Further improve separation of LootrunCommand

kristofbolyai committed:
Move static code to top of file

Incompleteusern commented on common/src/main/java/com/wynntils/commands/LootrunCommand.java:
```diff
@@ -0,0 +1,379 @@
+    @Override
+    public void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        LiteralCommandNode<CommandSourceStack> node = dispatcher.register(Commands.literal("lootrun")
+                .then(Commands.literal("load")
```
it might be nice to static import `Commands.literal` at this point, but debatable

Incompleteusern commented on common/src/main/java/com/wynntils/core/features/FeatureRegistry.java:
```diff
@@ -32,6 +32,15 @@ public static void registerFeature(Feature feature) {
         FEATURES.add(feature);
     }
 
+    public static void unregisterFeature(Feature feature) {
```
tbh when I made this it wasn't designed such that features can be unregistered, as any behavior with api dependencies or requirements could be handled by enabling/disabling instead.

Incompleteusern commented on common/src/main/java/com/wynntils/core/features/FeatureRegistry.java:
```diff
@@ -32,6 +32,15 @@ public static void registerFeature(Feature feature) {
+        }
+
+        FEATURES.remove(feature);
+        WynntilsMod.getEventBus().unregister(feature);
```
And I am pretty sure feature registration/unregistration is done by the feature code itself, but it could be debated to have it automatically done too

kristofbolyai replied:
Removed this code, I misunderstood how features work.

Incompleteusern commented on common/src/main/java/com/wynntils/mc/mixin/MultiPlayerGameModeMixin.java:
```diff
@@ -30,4 +37,14 @@ private void handleInventoryMouseClickPre(
 
         WynntilsMod.getEventBus().post(new ContainerClickEvent(containerId, slotId, itemStack, clickType, slotId));
```
it can be debated to make this an EventFactory method

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+
+    public static void unregisterFromEventBus() {
+        FeatureRegistry.unregisterFeature(lootrunFeatureInstance);
+    }
```
Like earlier comments, this is designed to be done by the feature itself, and should instead have the feature be enabled/disabled, unless that is changed.

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+            int color,
+            MultiBufferSource.BufferSource source,
+            long chunkLong) {
+        List<Pair<Vec3, Component>> notes = lootrun.notes().get(chunkLong);
```
A problem I've always had with lootrun code is the excessive use of `Pair`, imo `Pair<Vec3, Component>` can be replaced with a `Note` class, which would allow easier method names

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+            MultiBufferSource.BufferSource source,
+            Long2ObjectMap<List<List<ColoredPoint>>> points,
+            long chunkLong) {
+        List<List<ColoredPoint>> locations = points.get(chunkLong);
```
`Long2ObjectMap<List<List<ColoredPoint>>>  ` makes it exceedingly difficult to understand what is going on, please try to make it easier to understand that its a chunk map of list of paths

Incompleteusern replied:
`List<ColoredPoint>` and `List<Vec3>` could simply be a ColoredPath and Path class and hold relevant methods too

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+                    for (BlockPos blockInArea : blocks) {
+                        BlockState blockStateInArea = level.getBlockState(blockInArea);
+                        if (blockStateInArea.is(Blocks.BARRIER)) {
+                            barrierInArea = true;
```
`barrierInArea` check is a bit iffy, given that there exists barriers in use in SE

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+
+    public static class RecordingInformation {
+        private Vec3 lastLocation;
+        private BlockPos lastChest;
```
where is `lastLocation` and `lastChest` used?

kristofbolyai replied:
They are used in recording. `LootrunUtils#729 and 727`

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+        return 1;
+    }
+
+    public static LootrunSaveResult saveLootrun(String name) {
```
imo this could be a method of `LootrunUncompiled`, and on that note I feel as though `LootrunInstance` and `LootrunUncompiled` could be merged into one class, and that `RecordingInformation`'s `dirty` could also be situated there

kristofbolyai replied:
I would not merge the two, but I agree that methods can be moved.

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+
+        for (int i = 0; i < Math.pow(renderDistance, 2); i++) {
+            int x = i % renderDistance + origin.x - (renderDistance / 2);
+            int z = i / renderDistance + origin.z - (renderDistance / 2);
```
imo better off as 2 for loops

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+            int x = i % renderDistance + origin.x - (renderDistance / 2);
+            int z = i / renderDistance + origin.z - (renderDistance / 2);
+            ChunkPos chunk = new ChunkPos(x, z);
+            if (McUtils.mc().level == null || !McUtils.mc().level.hasChunk(chunk.x, chunk.z)) {
```
`McUtils.mc().level == null` should only be done once at beginning of rendering

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+
+                lastBlockPos = blockPos;
+
+                if (!pauseDraw) {
```
`pauseDraw` seems to be exactly the same as `disablePoint`, or at least functions the same

kristofbolyai replied:
`pauseDraw` is reset for every point, as it is declared inside the for loop, `disablePoint` is not, it is consistent through rendering.

Incompleteusern replied:
yeah but when `pauseDraw ` is used here its set to the same state as `disablePoint`

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+                boolean pauseDraw = false;
+                BlockPos blockPos = new BlockPos(point.vec3());
+
+                if (blockPos.equals(lastBlockPos)) { // Do not recalculate block validness
```
block validity should be moved to a separate method, `barrierInArea` is unnecessary beyond making it invalid to render and could be merged with `validBlock `

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -0,0 +1,814 @@
+            Matrix4f lastMatrix = poseStack.last().pose();
+            boolean sourceBatchEnded = false;
+
+            List<ColoredPoint> toRender = new ArrayList<>();
```
`toRender` seems to be slightly convoluted in its implementation, and seems to be a way of storing positions that are in the air or equal to last block position?? for later rendering.

kristofbolyai replied:
Points are collected every tick, so they happen to be the same block a lot of the times. That's why the caching exists, so it does not recalculate block validness.

kristofbolyai committed:
Extract level

kristofbolyai committed:
Refactor Features and refactor rendering code, fix issues with it

kristofbolyai committed:
Add static imports

kristofbolyai committed:
Add EventFactory method for mixin

kristofbolyai committed:
Add Note record

kristofbolyai committed:
Move saveLootrun

kristofbolyai committed:
Move saveLootrun

kristofbolyai committed:
Merge branch 'lootrun' of https://github.com/kristofbolyai/Artemis into lootrun

magicus reviewed (dismissed):
Looks basically okay now. It can surely be polished quite a lot more. If you feel like doing it, feel free to! But I'm happy enough, right now. Structure is good enough. Rendering is better. (But still the points rendering could be straightened out a bit more).

I'm not sure I understood why you made the FeatureRegistry a Map though? Couldn't see where you were using the class key.

kristofbolyai commented:
> Looks basically okay now. It can surely be polished quite a lot more. If you feel like doing it, feel free to! But I'm happy enough, right now. Structure is good enough. Rendering is better. (But still the points rendering could be straightened out a bit more).
> 
> 
> 
> I'm not sure I understood why you made the FeatureRegistry a Map though? Couldn't see where you were using the class key.

I will go ahead and finish all the stuff Incompleteusern suggested.

Also, the map is used in LootrunUtils, to enable and disable the feature.

kristofbolyai committed:
Extract block validness function

kristofbolyai committed:
Further refactor RenderPoints

kristofbolyai committed:
More refactors to variables

kristofbolyai committed:
Use the correct render type

magicus reviewed (dismissed):
As always, there's a lot of tweaking that still could be done. But I'm okay with accepting this into the code as it is, and we can continue working out remaining issues later.

(My main beef with this PR right now is the FeatureRegistry changes. We're making a change that does not really fit, just to be able to enable/disable a single feature from the outside. There's something fishy about this, and I don't like it. But it might be that we need to develop a better framework for handling non-feature-tied functionality first to fix this properly.)

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -185,29 +185,29 @@ private static void renderChests(
-                        toRender.add(point);
+                    if (!toRender.points().isEmpty()) {
+                        toRender.points().add(point);
                     }
```
What is `toRender`, and why does the above check and lines of code exist? Maybe @coehlrich knows.

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -284,6 +266,23 @@ private static void renderPoints(
+        for (BlockPos blockInArea : blocks) {
+            BlockState blockStateInArea = level.getBlockState(blockInArea);
+            if (blockStateInArea.is(Blocks.BARRIER)) {
+                state = BlockValidness.HAS_BARRIER;
```
should be an auto return,

kristofbolyai replied:
No, there could be a vaild block when there is a barrier.

Incompleteusern replied:
doesn't this create conflicts based off if the valid block or the barrier occurs first?

kristofbolyai replied:
I followed what the original code did. I think if we have a valid block, we don't need to care about barriers, but we still search for barriers in case there is no valid block, as the rendering process is different for all outcomes

kristofbolyai replied:
Original code: 
```java
boolean barrier = false;
boolean validBlock = false;
for (BlockPos blockInArea : blocks) {
    BlockState blockStateInArea = level.getBlockState(blockInArea);
    if (blockStateInArea.is(Blocks.BARRIER)) {
        // System.out.println("barrier");
        barrier = true;
    } else if (blockStateInArea.getCollisionShape(level, blockInArea) != null) {
        validBlock = true;
    }
}

 if (validBlock) {
      disable = false;
      if (disabled) {
          consumer = source.getBuffer(RenderType.lineStrip());
          disabled = false;
      }
      for (Point location : toRender) {
          Vec3 rawLocation = location.vec3();
          int pathColor = location.color();
          consumer.vertex(last, (float) rawLocation.x, (float) rawLocation.y, (float)
                          rawLocation.z)
                  .color(pathColor)
                  .normal(0, 0, 1)
                  .endVertex();
      }
      toRender.clear();
  } else if (barrier) {
      disable = true;
      pauseDraw = true;
      toRender.clear();
  } else if (disable) {
      pauseDraw = true;
  } else {
      toRender.add(loc);
      continue;
    }
}
                                    ```

Incompleteusern replied:
ah

Incompleteusern commented on common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java:
```diff
@@ -284,6 +266,23 @@ private static void renderPoints(
+            }
+        }
+
+        return state;
```
this should just be inlined as `BlockValidness.INVALID`

kristofbolyai replied:
If we don't auto return HAS_BARRIER it shouldn't.

Incompleteusern replied:
ah

DevScyu commented on common/src/main/java/com/wynntils/features/LootrunFeature.java:
```diff
@@ -0,0 +1,72 @@
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class LootrunFeature extends Feature {
```
Needs to follow the new FeatureBase format introduced in #60

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai committed:
Merge remote-tracking branch 'origin/lootrun' into lootrun

kristofbolyai committed:
Do not use wildcard imports

kristofbolyai committed:
Fix imports

Incompleteusern commented:
To me its fine for now, until we get a proper way of doing things to replace the hacks

kristofbolyai committed:
Merge branch 'main' into lootrun

kristofbolyai commented:
Need a new review 😐
